accepted/2024/support-nuget-authentication-plugins-dotnet-tools.md

```diff
@@ -0,0 +1,210 @@
+
+### Specify the the authentication plugin in NuGet.Config file
+
+To simplify the authentication process with private NuGet feeds, customers can specify the authentication plugins installed as .NET Tools in the `packagesourceCredentials` section of the NuGet.Config file.
+
+Here is an example of how to configure the NuGet.Config file:
+
+```xml
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <packageSources>
+    <add key="ExamplePrivateRepo" value="https://example.com/nuget" />
+    <!-- Add other package sources as needed -->
+  </packageSources>
+
+  <packageSourceCredentials>
+    <ExamplePrivateRepo>
+      <add key="ToolName" value="Microsoft.CredentialProvider" />
+      <!-- The value here is the .NET tool name installed globally or in a specific tool-path -->
+    </ExamplePrivateRepo>
+    <!-- Add credentials for other package sources as needed -->
+  </packageSourceCredentials>
+</configuration>
+```
+
+**Advantages:**
+- This approach explicitly configures the intent to use the .NET Tool as a plugin for authentication in the NuGet.Config file itself.
+- Customers can install the plugins as global .NET Tools, eliminating the need to specify a custom location based on the platform.
+
+**Disadvantages:**
+- Configuring the setting per feed can be cumbersome if multiple private feeds can use the same .NET tool for authentication.
+
+An alternative approach to address this disadvantage is to declare the plugins explicitly outside of the package source sections. This approach allows customers to specify the dependency only once, without configuring it per private feed as discussed above.
+
+```xml
+<?xml version="1.0" encoding="utf-8"?>
+<configuration>
+  <DotNetToolPlugins>
+    <Authentication>
+      <plugin id="CredentialProvider.Microsoft" />
+      <plugin id="CredentialProvider.GitHub" />
+      <!-- Allows specifying multiple authentication plugins with additional details -->
+    </Authentication>
+  <!-- This format provides flexibility for future plugin types and detailed configurations -->
+  </DotNetToolPlugins>
+</configuration>
+```
+
+However, the NuGet Client plugins functionality is a global setting. This means that the plugins are discovered based on the platform and target framework, as discussed earlier, and there is currently no way to configure the dependent plugins via NuGet settings per repository. It could be a future possibility to provide users with an option to manage their plugins per repository instead of loading all the available plugins. However, given the limited number of plugin implementations currently available, this is not a problem that needs to be solved at this point in time, in my understanding.
+
+## Prior Art
```

Comment:
Are NuGet credential helpers loaded in-proc? If so, why? Wouldn't a stdin/stdout protocol like [Git Credential Helpers](https://git-scm.com/docs/gitcredentials) or [Docker Credential Helpers](https://github.com/docker/docker-credential-helpers) be easier to manage from a dependency management perspective (no plugin/assembly loading issues) as well as from an ecosystem perspective (folks from any ecosystem could make a helper, not just .NET)?

Reply:
It's a separate process: https://learn.microsoft.com/en-us/nuget/reference/extensibility/nuget-cross-platform-plugins.

Reply:
Thanks @nkolev92!

Reply:
Part of the problem is that for `.NET Core` scenarios, NuGet is only looking for `.dll` files and invoking it explicitly with `dotnet ...\x.dll`. I agree that the cred provider shouldn't have to be written for .NET.

Reply:
side notes about Google's credential helpers noted above: the npm helper is like Azure Artifacts's vsts-npm-auth, which just writes the credentials into a .npmrc, and the maven helper appears to essentially swap out the HTTP request layer. Both are fundamentally different from what we're doing here. Cargo's credential providers are much more like NuGet's credential providers, and are configured in Cargo's config with a list of things to try in turn. NuGet's credential providers are partially inspired by Git's credential providers, which I believe has a single setting for what credential provider to use. Python tools use a slightly different approach, using the `keyring` package to store credentials, and that uses a convention-based mechanism to discover Python packages in the environment that provide credential providers.

Reply:
There's only extra complexity on the plugin author side if the plugin author chooses not to use .NET. Implementations that do use .NET can still use the NuGet libraries as before. The largest use-case I can see for non-.net would be _extremely simple_ plugins for e.g. CI providers, where instead of fully supporting the whole full-duplex asynchronous protocol, it can be approximated well enough through a synchronous, half-duplex implementation. There's nothing in the protocol that _requires_ that plugins be implemented in .NET.

Reply:
Yup, I agree. 
We just haven't heard any such feedback before, and I'm not sure there's a need for that at this point.

Reply:
The prior art mentioned in this thread refers to credential providers in various tools/languages. Given that NuGet already has a plugin model to support credential providers, and this specification merely proposes a new deployment and runtime discovery mechanism (using .NET tools), I didn't update the spec to include all these details.

The reason I mentioned the `AWS CodeArtifact NuGet Credential Provider` is because it has already been published as a .NET tool, which is within the scope of this specification.

IMHO, we should consider supporting non-.NET credential providers as a separate effort. Given the limited number of plugin implementations currently available, this is not a problem that needs to be solved at this point in time, in my understanding. Please correct me if I am missing something. Perhaps @phil-hodgson, @JohnSchmeichel, and @jmyersmsft have more information on how NuGet's support for non-.NET credential providers would benefit the Azure Artifacts credential provider.

Reply:
> this specification merely proposes a new deployment and runtime discovery mechanism (using .NET tools), I didn't update the spec to include all these details.

My question is what are the discovery and installation mechanisms of credential providers for these non-NuGet package managers. Do they have prior art we can learn from about how to discover and install plugins, or is NuGet a leader in this space?

If all or most of them are using a specific technique then I think there may be value in using the same technique in NuGet, rather than designing our own. If there's a defacto standard that we're not following, I think we should have a justification why.

> IMHO, we should consider supporting non-.NET credential providers as a separate effort.

This spec is already changing how NuGet scans for plugins, although in a minor way. However, if we can design it in a way that Just Works for cred providers not written in .NET, then it reduces potential future efforts. If this spec proposes a design that is known to be incompatible with cred providers written in other languages, then I think we should justify why. If it's not known if this design is compatible or incompatible with cred providers written in other languages, I think the spec is incomplete.

If dropping a shell script or a Go or Rust binary in the directory that this spec is proposing causes `dotnet tool list --tools-path`, or `dotnet tool update` to error out, then it would be good to know. One reason is because if we get bug reports from customers trying to update their cred provider package, we may know why before a potentially difficult investigation. A second reason is that we can consider if there's any EASY alternative. If there isn't we can make an educated decision to either change the design, or accept that future work will be needed if we need to handle the scenario (and possibly customer bug reports). But I feel more comfortable making an educated decision, rather than "oops, we didn't consider other scenarios when designing the feature".

But, if there is an "easy" alternative that is more flexible, then designing it that way now reduces potential future work. For example, if we can easily re-use the .NET SDK code that creates an executable for a .NET tool, then we can consider a custom package layout, similar to how MSBuild SDK packages and .NET Tool packages and Workload packages all use custom layouts compared to NuGet's original Libraries package type layout. Then we can allow non-.NET cred providers to ship in the same way as .NET cred providers, and we just create the entry point executable for .NET cred providers.

On the other hand, if putting shell scripts and other binaries in the new plugin directory doesn't cause problems with dotnet tools, then we can document that cred providers can be written in any language, but they just can't distribute via `dotnet tool install` or discover via `dotnet tool search`.

Or as other people have written in other comment threads, if we have a specific executable naming convention and we scan all directories in PATH for files that match this naming convention, then .NET tools can be installed via `dotnet tool install --global`, and non-.NET cred providers can use winget, chocolatey, or whatever else they want.

Or as proposed elsewhere, if nuget.config tells nuget to "use a credential provider whose executable name is Contoso-CredProvider", then we don't even need to scan PATH, just execute it as configured (although we'd have to figure out what this means in VS). This way non-.NET cred providers can Just Work, even if they lose out to .NET cred providers discoverability via `dotnet tool search CredProvider`, but that could also be considered a competative advantage of cred providers that are written in .NET.

It's not my intention to cause scope creep, but I see it as a way to design extensibility to enable non-core scenarios to avoid future work. If we understand the limitations of different designs we can trade off flexibility vs lower implementation cost. If we don't understand the limitations of different designs, I feel like I can't make an educated decision on which is best design to approve and implement.

Reply:
> This spec is already changing how NuGet scans for plugins, although in a minor way. However, if we can design it in a way that Just Works for cred providers not written in .NET, then it reduces potential future efforts. If this spec proposes a design that is known to be incompatible with cred providers written in other languages, then I think we should justify why. If it's not known if this design is compatible or incompatible with cred providers written in other languages, I think the spec is incomplete.

I had a quick chat with @dtivel regarding the possibility of supporting non-.NET plugins. I was informed that the NuGet Client plugin code must be updated to utilize a standard RPC mechanism, such as StreamJsonRpc, to support that use case. I will add this use case to the future possibilities section.

Reply:
> My question is what are the discovery and installation mechanisms of credential providers for these non-NuGet package managers. Do they have prior art we can learn from about how to discover and install plugins, or is NuGet a leader in this space?

@jmyersmsft added some information about the non-NuGet credential providers in https://github.com/NuGet/Home/pull/13242/files/47e844a79558e412d2d7114ae0b7e13ef5d30093#discussion_r1492866045. (I am reading the links again incase if I find something that helps us).

Reply:
> I was informed that the NuGet Client plugin code must be updated to utilize a standard RPC mechanism, such as StreamJsonRpc, to support that use case.

"must" is an overly strong characterization.

What I meant is that NuGet's IPC is home-grown, and it would be to the advantage of anyone implementing a plugin if the IPC were more standards-based.  Then, a plugin author could reuse an existing implementation (for another language, perhaps).

I don't think there's anything blocking anyone from implementing a client in a non-.NET language; however, there would be an additional cost of reimplementing the plugin protocol, which is a deterrent for writing a non-.NET plugin.